### DISCUSSION TITLE
Tips for using Eclipse IDE for Glassfish source code development?

### DISCUSSION THREAD

**Discussion by escay** [eclipse-ee4j/glassfish#24854] [Q&A]
My current approach for using Eclipse IDE for Glassfish source code development is as follows:

Install java, maven and build all of Glassfish
- download and install JDK (jdk-17.0.8.1_1-openjdk-adoptium) - https://adoptium.net/temurin/releases/?version=17
- download and install Maven (apache-maven-3.9.6) - https://maven.apache.org/download.cgi
- checkout Glassfish code and build everything using 'mvn clean install', ensure it succeeds

Install Eclipse
- download and install Eclipse, Java Developers version (eclipse-java-2023-12-R-win32-x86_64) - https://www.eclipse.org/downloads/packages/
- edit eclipse.ini, increase memory: -Xms1024m -Xmx8192m (not sure yet what would be a good start value)
- Start Eclipse
  - Update all plugins
  - Configure local maven instead of embedded, and point to personal /conf/settings.xml
  - Configure local JDK instead of embedded
  - Configure code formatting: https://github.com/eclipse-ee4j/ee4j/tree/master/codestyle
- Restart Eclipse to have new Maven and JDK active
- Import maven modules using m2eclipse
  - File -> Import -> Existing Maven Projects
  - Select all '350+' glassfish modules
  
This works when opening a few maven modules at a time / being selective in which modules I open, but importing all of them does not work for me. Which prevents me from looking around all code in a simple manner.

Main issue is: a few build path problems and quite a lot of m2eclipse 'mojo' errors.

Example:
![image](https://github.com/eclipse-ee4j/glassfish/assets/1008888/7c864406-b82d-4a53-a2f6-e030a0bdbfcb)

Are there any tips for using Eclipse for Glassfish source code development?

**Comment by dmatej** [comment]
This is a bit problematic, on one side we slowly improve the GlassFish build, on the second side Eclipse IDE improves too, especially the m2e plugin. The latest m2e improved again, while previous version had issues with hiding "derived". Basically you need the **lifecycle file** and configure it the way you see as the right for your case. When you press CTRL+1 on an error, Eclipse gives some suggestions.

There are also inactive parts with tests for GlassFish 2, which we still did not delete. Until GlassFish 6 especially clusters did not work properly while they worked well in Sun GlassFish Enterprise server; these tests could be used as an inspiration to write new tests using TestContainers (I have such plan and I already have some locally). On the other hand we could simply delete them and just mention some link/hash/tag/branch where it happened. This step could resolve part of those errors and it could make git clone bit faster too.

What I usually do is that I am importing **all projects in the tree**, but first I am building them with `mvn clean install -T4C -Pfastest` - Eclipse IDE then reacts better, because there are no generated classes missing, etc.

But having the discussion on this theme might be really useful :-)

**Reply by escay** [reply]
The following lifecycle-mapping-metadata.xml ignores the m2eclipse mojo errors in my environment.
[lifecycle-mapping-metadata.xml.txt](https://github.com/eclipse-ee4j/glassfish/files/14572134/lifecycle-mapping-metadata.xml.txt)
I could not solve all of them using CRTL+1, had to add some manually.

**Reply by escay** [reply]
@dmatej is there a way to easily recognize the "inactive parts with tests for GlassFish 2" modules?
If I select "all projects in the tree" I think Eclipse imports all folders with a pom.xml in it.

**Reply by dmatej** [reply]
If I am not wrong, Eclipse offers just projects under the "aggregator pom", while those GF2 tests are mostly Ant based and stand aside.

**Comment by arjantijms** [comment]
Note that GlassFish also has a special debug project that includes all sources:

https://github.com/eclipse-ee4j/glassfish/tree/master/appserver/featuresets/debug

**Comment by anbusampath** [comment]
> importing all of them does not work for me. Which prevents me from looking around all code in a simple manner.

When importing all in one go, leads to lot of errors in Eclipse.
<img width="375" alt="image" src="https://github.com/user-attachments/assets/67fae4c6-e8ea-4a21-9891-beab6a44962b" />

> What I usually do is that I am importing **all projects in the tree**, but first I am building them with `mvn clean install -T4C -Pfastest` - Eclipse IDE then reacts better, because there are no generated classes missing, etc.

I will try importing this way.

**Reply by dmatej** [reply]
In current Eclipse IDE I seem weird paths added to classpath - it is probably a bug in Eclipse IDE or some issue caused by our maven settings. I did not have time to resolve this yet as I am switching workspaces and branches too fast now. If you would have some idea about it, I would appreciate it.
